File: MainForm.vb
Public Class MainForm

	Protected Friend t As Thread
	Protected WithEvents piCalc As CalculatePi
	''' <summary>CR &amp; LF = Windows New Line</summary>
	Public Const CrLf As String = Chr(13) & Chr(10)

	Private Sub Form1_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
		' Title information
		Me.Text = "Pi Calculator (" & IntToCPUCount(1, True) & ")"
		' CPU Information
		Dim q As New ObjectQuery("SELECT * FROM Win32_Processor")
		Dim p As New ManagementObjectSearcher(q)
		Dim i, i2 As Long ' Long Is Int64
		Dim s As String = Nothing
		For Each cpu As ManagementObject In p.Get()	' there could be more than one processor, like a server?
			i += CInt(cpu("NumberOfCores"))	' counts cores ' reminds me when I used i += 1, "why doesn't VB have i++ like c++?"
			i2 += CInt(cpu("NumberOfLogicalProcessors")) ' counts threads
			If s Is Nothing Then ' CPU name
				s = cpu("Name").ToString.Replace("   ", "") & ", " & (CInt(cpu("CurrentVoltage")) / 10) & "v"
			End If
		Next
		lblCPU.Text = s & ", " & i & " cores, " & i2 & " threads"
		' OS Information
		q = New ObjectQuery("SELECT * FROM Win32_OperatingSystem")
		p = New ManagementObjectSearcher(q)
		For Each os As ManagementObject In p.Get()
			lblOS.Text = CStr(os("Caption")) & " " & CStr(os("Version")) & " " & CStr(os("OSArchitecture"))
			Exit For ' stupid collection doesn't let me do p.Get()[0]
		Next
		' Memory Information
		q = New ObjectQuery("SELECT * FROM Win32_PhysicalMemory")
		p = New ManagementObjectSearcher(q)
		i = 0
		s = Nothing
		For Each mm As ManagementObject In p.Get()
			i += CLng(mm("Capacity")) ' add the memory module's capacity
			If s Is Nothing Then ' describe the memory
				Select Case CShort(mm("MemoryType")) ' Covers DDR and DDR2, others and assumes DDR3
					Case 1 To 19 ' not DDR memory
						s = "Unknown"
					Case 20
						s = "DDR"
					Case 21
						s = "DDR2"
					Case Else ' Assume unknown memory (Case 0) to be DDR3
						s = "DDR3"
				End Select
				s &= "-" & CStr(mm("Speed"))
			End If
		Next
		Dim sf As Char() = {"K"c, "M"c, "G"c}
		i2 = 0
		Do Until (i / (1024 ^ i2) < 1024) Or i2 > sf.Length
			i2 += 1
		Loop
		lblMemory.Text = s & " " & (i / (1024 ^ i2)) & If(i2 > 0, sf(CInt(i2 - 1)), "") & "B"
		i2 = 0 ' now counting elements to pop
		Select Case i ' bytes of memory
			Case Is < 268435456	' You should have at least 256 MB of RAM
				Me.Close() ' should not use this program
			Case 268435456 To 536870912	' 256MB to 512MB
				i2 = 5 ' not suitable for more than 16M
		End Select
		For i = 1 To i2
			MsgBox("rem")
			cmbPrecision.Items.RemoveAt(cmbPrecision.Items.Count - 1)
		Next
		If i2 > 0 Then ' popped elements, update maximum
			maxValChanged(Me, Nothing)
		End If
		' ComboBox Initalization - Visual Studio doesn't have these important properties available for ComboBox controls
		cmbPrecision.SelectedIndex = 7
		cmbDScale.SelectedIndex = 0
		cmbBuffer.SelectedIndex = 1
		' Store progress size difference
		progress.Tag = Me.Width - progress.Width
	End Sub

	''' <summary>Converts an integer to the number of threads (CPU thread count: 0, single, double, triple, quad, five...)</summary>
	''' <param name="i">The integer to convert</param>
	''' <param name="capital">If the result should be capital</param>
	Function IntToCPUCount(ByVal i As Integer, Optional ByVal capital As Boolean = False) As String
		Dim result As String = i.ToString
		Select Case i
			Case 1 : result = "single"
			Case 2 : result = "double"
			Case 3 : result = "triple"
			Case 4 : result = "quad"
			Case 5 : result = "five"
			Case 6 : result = "six"
			Case 7 : result = "seven"
			Case 8 : result = "eight"
			Case 9 : result = "nine"
			Case 10 : result = "ten"
			Case 11 : result = "eleven"
			Case 12 : result = "twelve"
		End Select
		result &= " thread" & If(i = 1, "", "s")
		If capital Then result = StrConv(result, VbStrConv.ProperCase) Else result = result.ToLower
		Return (result)
	End Function

	' Fix for the size of the progressbar in the status bar
	Private Sub MainForm_SizeChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles Me.SizeChanged
		If IsNumeric(progress.Tag) Then progress.Size = New Drawing.Size(Me.Width - CInt(progress.Tag), progress.Height)
	End Sub

	Private Sub copyTextBtn_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles copyTextBtn.Click
		' Copy the static label text
		Clipboard.SetText(CType(CType(CType(sender, ToolStripMenuItem).Owner, ContextMenuStrip).SourceControl, Control).Text)
	End Sub

	Private Sub numPrecision_ValueChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles numPrecision.ValueChanged
		' IndexOf Note: [compare not] 0 because -1 is not there, 0 is first char
		If numPrecision.Value Mod KprecisionP(2) = 0 Then
			' M
			Dim j As UInteger = CUInt(numPrecision.Value / KprecisionP(2))
			For Each i As String In cmbPrecision.Items
				If i.IndexOf("M"c) > 0 And i.Replace("M", "") = CStr(j) Then
					cmbPrecision.SelectedItem = i
					Exit Sub
				End If
			Next
		End If
		If numPrecision.Value Mod KprecisionP() = 0 Then
			' K
			Dim j As UInteger = CUInt(numPrecision.Value / KprecisionP())
			For Each i As String In cmbPrecision.Items
				If i.IndexOf("K"c) > 0 And i.Replace("K", "") = CStr(j) Then
					cmbPrecision.SelectedItem = i
					Exit Sub
				End If
			Next
		End If
		For Each i As String In cmbPrecision.Items
			If i.IndexOf("K"c) < 0 And i.IndexOf("M"c) < 0 And i = CStr(numPrecision.Value) Then
				cmbPrecision.SelectedItem = i
				Exit Sub
			End If
		Next
		cmbPrecision.SelectedItem = "?"
	End Sub

	Private Sub precisionComboChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cmbPrecision.SelectedIndexChanged, cmbDScale.SelectedIndexChanged
		If cmbPrecision.SelectedIndex = 0 Then
			numPrecision_ValueChanged(sender, New EventArgs)
		ElseIf CStr(cmbPrecision.SelectedItem).IndexOf("M"c) > 0 Then
			numPrecision.Value = CInt(CStr(cmbPrecision.SelectedItem).Replace("M", "")) * KprecisionP(2)
		ElseIf CStr(cmbPrecision.SelectedItem).IndexOf("K"c) > 0 Then
			numPrecision.Value = CInt(CStr(cmbPrecision.SelectedItem).Replace("K", "")) * KprecisionP()
		Else
			numPrecision.Value = CInt(cmbPrecision.SelectedItem)
		End If
	End Sub

	Public Function KprecisionP(Optional ByVal pow As Byte = 1) As ULong
		Return CULng(Math.Pow(If(cmbDScale.SelectedIndex = 0, 1024, 1000), pow))
	End Function

	Private Sub maxValChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cmbDScale.SelectedIndexChanged
		Dim l As String = CStr(cmbPrecision.Items(cmbPrecision.Items.Count - 1))
		If l.IndexOf("M"c) > 0 Then
			l = l.Replace("M", "")
			l = CStr(CInt(l) * KprecisionP(2))
		ElseIf l.IndexOf("K"c) > 0 Then
			l = l.Replace("K", "")
			l = CStr(CInt(l) * KprecisionP())
		End If
		numPrecision.Maximum = CDec(l)
	End Sub

	Private Sub btnGo_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnGo.Click
		' update GUI to allow stop
		btnGo.Enabled = False
		btnStop.Enabled = True
		' lock GUI
		numPrecision.Enabled = False
		cmbDScale.Enabled = False
		cmbPrecision.Enabled = False
		' say processing
		txtResult.Text = "Processing"
		' update progress bar
		progress.Value = 0
		progress.Maximum = 100
		progressText.Text = "0%"
		' start thread
		piCalc = New CalculatePi(CInt(numPrecision.Value))
		piCalc.startTicks = Now.Ticks
		' create the thread
		t = New Thread(AddressOf piCalc.Process)
		t.Start()
		t.Name = "Pi Calculator Calculation Thread"
	End Sub

	Public Delegate Sub controlEventInvoker(ByVal sender As System.Object, ByVal e As System.EventArgs)

	Protected Friend Sub stopThread(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnStop.Click, piCalc.onComplete, Me.FormClosing
		If btnGo.Enabled Then Exit Sub
		If Me.InvokeRequired Then
			Me.Invoke(New controlEventInvoker(AddressOf stopThread), sender, e)
			Exit Sub
		End If
		' update GUI to allow start
		btnGo.Enabled = True
		btnStop.Enabled = False
		' unlock GUI
		numPrecision.Enabled = True
		cmbDScale.Enabled = True
		cmbPrecision.Enabled = True
		' fill progress bar
		progress.Value = 1
		progress.Maximum = 1
		' update progress text
		progressText.Text = "Idle"
		' delete threads
		If t.IsAlive Then t.Abort() ' stop thread before delete
		t = Nothing	' delete the instance, .NET will reclaim the memory for me
		' retrieve data and delete piCalc
		Dim r As CalculatePi.TimedResult = piCalc.ResultDataNoDelete(If(cmbBuffer.SelectedIndex > 0, CalculatePi.ResultType.First2K, CalculatePi.ResultType.BufferOnly))
		txtResult.Text = If(sender Is btnStop, "Calculation was stopped" & CrLf, "") & r.s
		' ticks = 100-nanoseconds
		r.timeDiff = New TimeSpan(Now.Ticks - r.timeStart)
		lblDisplay.Text = r.timeDiff.Seconds.ToString.PadLeft(2, "0"c) & "s " & r.timeDiff.Milliseconds.ToString.PadLeft(3, "0"c) & "ms " & _
		 CStr(Math.Round(r.timeDiff.Ticks * 10) Mod 1000).PadLeft(3, "0"c) & "µs"
		piCalc.diffTicks = New TimeSpan(Now.Ticks - piCalc.startTicks)
		lblCalc.Text = piCalc.diffTicks.Hours.ToString.PadLeft(2, "0"c) & "h " & piCalc.diffTicks.Minutes.ToString.PadLeft(2, "0"c) & "m " & _
		 piCalc.diffTicks.Seconds.ToString.PadLeft(2, "0"c) & "s " & piCalc.diffTicks.Milliseconds.ToString.PadLeft(3, "0"c) & "ms " & _
		 CStr(Math.Round(piCalc.diffTicks.Ticks * 10) Mod 1000).PadLeft(3, "0"c) & "µs"
		If cmbBuffer.SelectedIndex > 1 Then	' save to file
			If sender Is btnStop Then
				MsgBox("Calculation was stopped; not saving your calculation", MsgBoxStyle.Critical, "Cannot save file!")
			Else
				saveDialog.InitialDirectory = Environment.GetFolderPath(Environment.SpecialFolder.Desktop)
				If saveDialog.ShowDialog = Windows.Forms.DialogResult.OK Then
					Dim sPath As String = saveDialog.FileName
					If sPath.Length > 33 Then sPath = "..." & sPath.Substring(sPath.Length - 30)
					MsgBox(sPath, MsgBoxStyle.Information, "File saved to:")
					r = piCalc.ResultData(CalculatePi.ResultType.All)
					File.WriteAllText(saveDialog.FileName, r.s)
				End If
			End If
		End If
		piCalc = Nothing
	End Sub

	Public Delegate Sub oneParamTypeInvoker(Of paramType)(ByVal i As paramType)

	Protected Friend Sub calcProgress(ByVal p As Byte) Handles piCalc.onProgress
		If Me.InvokeRequired Then
			Me.Invoke(New oneParamTypeInvoker(Of Byte)(AddressOf calcProgress), p)
		End If
		progress.Value = p ' progress bar
		progressText.Text = p & "%"	' progress text
	End Sub

	Private Sub menuClose_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles menuClose.Click
		Me.Close()
	End Sub
End Class

File: tags/alpha4-fail/MainForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using System.Management;
using System.Threading;
using System.IO;

namespace Pi
{


	public partial class MainForm : Form
	{
		protected Thread t;
		protected CalculatePi calc;

		public MainForm()
		{
			InitializeComponent();
		}

		private void MainForm_Load(object sender, EventArgs e)
		{
			this.Text = "Pi Calculator (" + IntToCPUCount(1, true) + ")";
			// CPU Information
			ObjectQuery q = new ObjectQuery("SELECT * FROM Win32_Processor");
			ManagementObjectSearcher p = new ManagementObjectSearcher(q);
			long i = 0, i2 = 0;
			string s = "";
			foreach (ManagementObject cpu in p.Get()){
				i += (int)cpu["NumberOfCores"]; // counts cores
				i2 += (int)cpu["NumberOfLogicalProcessors"]; // counts threads
				if (s == "") s = cpu["Name"].ToString().Replace("   ", "") + ", " + ((UInt16)cpu["CurrentVoltage"] / 10) + "v";
			}
			lblCPU.Text = s + ", " + i + " cores, " + i2 + " threads";
			// OS Information
			q = new ObjectQuery("SELECT * FROM Win32_OperatingSystem");
			p = new ManagementObjectSearcher(q);
			ManagementObject os = (ManagementObject)p.Get().GetEnumerator().Current;
			lblOS.Text = os["Caption"] + " " + os["Version"] + " " + os["OSArchitecture"];
			// Memory Information
			q = new ObjectQuery("SELECT * FROM Win32_PhysicalMemory");
			p = new ManagementObjectSearcher(q);
			i = 0; s = "";
			foreach (ManagementObject mm in p.Get()) {
				i += (long)mm["Capacity"];
				if (s == "") {
					UInt16 mt = (UInt16)mm["MemoryType"];
					switch (mt) {
						case 20: s = "DDR"; break;
						case 21: s = "DDR2"; break;
						default:
							if(mt >= 1 && mt <= 19) s = "Unknown"; // non-DDR memory
							else s = "DDR3?"; //  Assume unknown memory (case 0, and 22+) to be DDR3
							break;
					}
				}
			}
			char[] sf = {'k', 'm', 'g'};
			i2 = 0;
			while ((i / (Math.Pow(1024, i2)) >= 1024) && sf.Length > i2) i2++;
			lblMemory.Text = s + " " + (i / Math.Pow(1024 , i2)) + (i2 > 0 ? sf[i2 - 1].ToString() : "") + "B";
			// Store progress size difference
			progress.Tag = this.Width - progress.Width;
		}

		private void MainForm_OnFormClosing(object sender, FormClosingEventArgs e){
			if (btnStop.Enabled){
				e.Cancel = true;
				btnStop_Click(sender, e);
			}
		}

		/// <summary>Converts an integer to the number of threads (CPU thread count: 0, single, double, triple, quad, five...)</summary>
		/// <param name="i">The integer to convert</param>
		/// <param name="capital">If the result should be capital</param>
		private string IntToCPUCount(int i, bool capital = false) { 
			string result = i.ToString();
			switch(i){
				case 1 : result = "single"; break;
				case 2: result = "double"; break;
				case 3: result = "triple"; break;
				case 4: result = "quad"; break;
				case 5: result = "five"; break;
				case 6: result = "six"; break;
				case 7: result = "seven"; break;
				case 8: result = "eight"; break;
				case 9: result = "nine"; break;
				case 10: result = "ten"; break;
				case 11: result = "eleven"; break;
				case 12: result = "twelve"; break;
			}
			result += " thread" + (i == 1 ? "": "s");
			System.Globalization.TextInfo txtinfo = System.Globalization.CultureInfo.CurrentCulture.TextInfo;
		    if(capital) result = txtinfo.ToTitleCase(result);
			else result = result.ToLower();
			return result;
		}

		// Fix for the size of the progressbar in the status bar
		private void MainForm_SizeChanged(object sender, EventArgs e){
			progress.Size = new System.Drawing.Size(this.Width - (int)progress.Tag, progress.Height);
		}

		private void copyTextBtn_Click(object sender, EventArgs e)
		{
			// Copy the static label text
			Clipboard.SetText(((ContextMenuStrip)((ToolStripMenuItem)sender).Owner).SourceControl.Text);
		}

		private void numPrecision_ValueChanged(object sender, EventArgs e)
		{
			if (numPrecision.Value % KprecisionP(2) == 0) { 
				// M
				string j = Convert.ToString((uint)(numPrecision.Value / KprecisionP(2)));
				foreach (string i in cmbPrecision.Items) if (i.IndexOf('M') > 0 && i.Replace("M", "") == j){
					cmbPrecision.SelectedItem = i;
					return;
				}
			}
			if (numPrecision.Value % KprecisionP() == 0) {
				string j = Convert.ToString((uint)(numPrecision.Value / KprecisionP()));
				foreach (string i in cmbPrecision.Items) if (i.IndexOf('K') > 0 && i.Replace("K", "") == j){
					cmbPrecision.SelectedItem = i;
					return;
				}
			}
			foreach (string i in cmbPrecision.Items) if (i.IndexOf('M') < 0 && i.IndexOf('K') < 0 && i == numPrecision.Value.ToString()){
				cmbPrecision.SelectedItem = i;
				return;
			}
			cmbPrecision.SelectedItem = "?";
		}

		private void precisionComboChanged(object sender, EventArgs e) // cmbPrecision.SelectedIndexChanged, cmbDScale.SelectedIndexChanged
		{
			if (cmbPrecision.SelectedIndex == 0) numPrecision_ValueChanged(sender, e);
			else if (cmbPrecision.SelectedItem.ToString().IndexOf('M') > 0)
				numPrecision.Value = Convert.ToInt32(cmbPrecision.SelectedItem.ToString().Replace("M", "")) * KprecisionP(2);
			else if (cmbPrecision.SelectedItem.ToString().IndexOf('K') > 0)
				numPrecision.Value = Convert.ToInt32(cmbPrecision.SelectedItem.ToString().Replace("K", "")) * KprecisionP();
			else numPrecision.Value = Convert.ToInt32(cmbPrecision.SelectedItem);
		}

		public int KprecisionP(byte pow = 1) {
			return (int)Math.Pow(cmbDScale.SelectedIndex == 0 ? 1024 : 1000, pow);
		}

		private void maxValChanged(object sender, EventArgs e) // cmbDScale.SelectedIndexChanged
		{
			string l = (string)cmbPrecision.Items[cmbPrecision.Items.Count - 1];
			if(l.IndexOf('M') > 0){
				l = l.Replace("M", "");
				l = (Convert.ToInt32(l) * KprecisionP(2)).ToString();
			}
			else if (l.IndexOf("K") > 0){
				l = l.Replace("K", "");
				l = (Convert.ToInt32(l) * KprecisionP()).ToString();
			}
			numPrecision.Maximum = Convert.ToDecimal(l);
			precisionComboChanged(sender, e);
		}

		private void btnGo_Click(object sender, EventArgs e)
		{
			// update GUI to allow stop
			btnGo.Enabled = false;
			btnStop.Enabled = true;
			// lock GUI
			numPrecision.Enabled = cmbDScale.Enabled = cmbPrecision.Enabled = false;
			// say processing
			txtResult.Text = "Processing";
			// update progress bar
			progress.Value = 0;
			progressText.Text = "0%";
			
			// create the thread
			calc = new CalculatePi((int)numPrecision.Value);
			calc.onProgress += new ByteHandler(calcProgress);
			calc.onComplete += new EventHandler(btnStop_Click);
			t = new Thread(calc.process);
			t.Name = "Pi Calculator Calculation Thread";
			// start thread
			t.Start();
		}

		private void btnStop_Click(object sender, EventArgs e)
		{
			if (btnGo.Enabled || t == null) return;
			if (InvokeRequired) {
				Invoke(new EventHandler(btnStop_Click), sender, e);
				return;
			}
			// update GUI to allow start
			btnGo.Enabled = true;
			btnStop.Enabled = false;
			// unlock GUI
			numPrecision.Enabled = cmbDScale.Enabled = cmbPrecision.Enabled = true;
			// fill progress bar
			progress.Value = 100;
			// update progress text
			progressText.Text = "Idle";
			// delete thread
			t.Abort();
			t = null;
			// calculation time
			TimeSpan timeDiff = new TimeSpan(DateTime.Now.Ticks - calc.startTime);
			lblCalc.Text = timeDiff.Hours.ToString().PadLeft(2, '0') + "h " + timeDiff.Minutes.ToString().PadLeft(2, '0') + "m " +
				timeDiff.Seconds.ToString().PadLeft(2, '0') + "s " + timeDiff.Milliseconds.ToString().PadLeft(3, '0') + "ms " +
				(Math.Round(timeDiff.Ticks * 10d) % 1000).ToString().PadLeft(3, '0') + "µs";
			// retrieve data
			CalculatePi.timedResult r = calc.ResultData(cmbBuffer.SelectedIndex > 0 ? CalculatePi.timedResult.resultType.First2K : CalculatePi.timedResult.resultType.BufferOnly);
			timeDiff = new TimeSpan(DateTime.Now.Ticks - r.timeStart);
			txtResult.Text = sender == btnStop ? "Calculation was stopped\r\n" : "";
			txtResult.Text += r.s;
			lblDisplay.Text = timeDiff.Seconds.ToString().PadLeft(2, '0') + "s " + timeDiff.Milliseconds.ToString().PadLeft(3, '0') + "ms " +
				(Math.Round(timeDiff.Ticks * 10d) % 1000).ToString().PadLeft(3, '0') + "µs";
			if (cmbBuffer.SelectedIndex > 1) { // save to file
				if (sender == btnStop)
					MessageBox.Show("Calculation was stopped; not saving your calculation", "Cannot save file!", MessageBoxButtons.OK, MessageBoxIcon.Error);
				else {
					SaveDialog.InitialDirectory = Environment.GetFolderPath(Environment.SpecialFolder.Desktop);
					if (SaveDialog.ShowDialog() == DialogResult.OK) {
						string sPath = SaveDialog.FileName;
						if (sPath.Length > 33) sPath = "..." + sPath.Substring(sPath.Length - 30);
						r = calc.ResultData(CalculatePi.timedResult.resultType.All);
						File.WriteAllText(SaveDialog.FileName, r.s);
						MessageBox.Show(sPath, "File saved to:", MessageBoxButtons.OK, MessageBoxIcon.Information);
					}
				}
			}
			// delete calculator
			calc = null;
			// check if abort needed
			if (sender == this) Close();
		}

		protected void calcProgress(byte p) {
			if (InvokeRequired) { Invoke(new ByteHandler(calcProgress), p); return; }
			progress.Value = p;
			progressText.Text = p.ToString().PadLeft(2, '0') + "%";
		}

		private void menuClose_Click(object sender, EventArgs e) { this.Close(); }

		public bool creditsShown = false;
		private void menuCredits_Click(object sender, EventArgs e)
		{
			if (creditsShown) return;
			Credits c = new Credits(this);
			c.Show();
		}
	}
}
